### ISSUE TITLE
[Pi4] performance dedicated thread

### ISSUE THREAD

**Issue by nemo93** [dosbox-staging/dosbox-staging#420]
Hello Guys, I'm still impressed by the energy and work you put in this repo. For sure it's appreciated! 

I recently got my hands on a Pi4 and it's a beast compared to Pi3! I gladly did another test run with my various games and I was disappointed by the outcome. 

There was little to no difference in terms of "speed" in some demanding games eg. the Screamer serie. Also the issue with Tomb Raider #241 is still there as well as UAKM #255 (DK seems to run ok though I need to double-check). I can't even get to the main menu in Tomb Raider and UAKM is slow like hell while displaying the main menu (it's like a slideshow). In general anytime there's a FMV or video intro etc the playback heavily stutter to the point it becomes a slideshow. Worse than that some games like Alone in the dark serie appear to run slower than on Pi3 (especially Jack in the dark episode).

**_Question_**: are there specific flags or optimizations to apply in order to compile Dosbox-staging on a Pi4? 

As always I've carefully followed the [build guide](https://github.com/dosbox-staging/dosbox-staging/blob/master/BUILD.md) to compile my version:
`./scripts/build.sh -c gcc -t release -m lto`

Pi4 is running on Buster with all packages up-to-date as of typing. Retropie 4.6 is there too if that matters.

I've also updated the `dosbox-staging-git.conf` file to be sure to have all parameters. Below is an extract of this conf I'm using with most games. 

**Note**: setting these `fullresolution = desktop / windowresolution = original / output = texture` is the only way to get a centered-fullscreen display.
```
[sdl]
fullscreen       = true
display          = 0
fullresolution   = desktop
windowresolution = original
output           = texture
texture_renderer = auto
capture_mouse    = seamless middlerelease
sensitivity      = 100
waitonerror      = true
priority         = higher,normal
mapperfile       = mapper-sdl2-git.map

[dosbox]
language = 
machine  = svga_s3
captures = capture
memsize  = 16

[render]
frameskip = 0
aspect    = true
scaler    = none
glshader  = sharp

[cpu]
core      = auto
cputype   = auto
cycles    = auto
cycleup   = 10
cycledown = 20
[etc.]
```

**Comment by kcgen** [comment]
Haven't picked up a Pi4 yet (so unfortunately can't follow along directly), but I too was hoping it would "unlock' the next tier of performance in dosbox gaming (say, in the ~60,000-cycle range)  versus the Pi3 (which is ~30,000 cycles). 

Let's check any low-hanging fruit:

1. The configure script should be enabling the dynrec core. Grep for `C_DYNREC` in your `config.h`.  If it's enabled, you will see: `#define C_DYNREC 1`.   If that's enabled, you can then force its use with: `core = dynamic` in your conf. 

1. On the Pi3, I also use `scaler = none`, but I also avoid the overhead of the shaders by using `glshader = none`, and then stretching up with either `texture_renderer = texturepp` or `texture_renderer = texturenb`.  Maybe that will help a bit.

1. CFLAGS (set by the build script) should be pretty close to optimal, as [we're using](https://github.com/dosbox-staging/dosbox-staging/blob/master/scripts/automator/build/gcc-linux_rpi4) the same one for the Pi4 [as RetroPie uses to build the Retroarch cores](https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/system.sh#L408).  That said, you can always tweak them and take measurements.

1. Another thing I do is aggressively control my cycle count on a per-game basis (each game has it's own local `dosbox.conf` just for little adjustments like this).  If you leave cycles to `auto`, atleast on the Pi3, it tends to overshoot and CPU intensive games can feel more bogged down than using a fixed cycles count.  

Start with the Quake 1 timedemo and a fixed cycle count of say 25,000.  On another system, SSH into your Pi4 and run time - and watch dosbox's CPU usage in `top`.  Keep ratchetting up cycles in say 5k increments, until `dosbox-staging` consumes around 90% or so CPU time.  This will also be around the edge when gameplay starts to feel a big choppier; in which case you can back off the cycles bit by bit until you found your max.

1. I also use `vgaonly` (early 90s VGA games) or `vesa_nolfb` (mid-90s or higher resolution games), unless a game specifically needs something else.

1. Another thing I do is disable all the sound devices and only enable those used by the game.  Most SB-based games work well with `sbpro2` instead of `sb16` (unless you notice their configuration program specifically wanting a High DMA or HDMA - then you know the game uses 16-bit samples).  I tend to use a sampling rates of 16000 for PC Speaker games, 22050 for older games, and 44100 for newer or CDDA games or 48000 for CDDA games w/ Opus files.  That's all pretty minor though.

This is all just shotgun advice though.. it might all be worthless.  Curious what you find!

**Comment by kcgen** [comment]
@nemo93, any luck?  

Was thinking about it, and a wiki page would be great if we can nail down recommended  settings on the Pi. Some might just be recommendations, but others might essentially be mandatory.

**Comment by nemo93** [comment]
hey @kcgen sorry I couldn't respond earlier... swamped with other things! I really appreciate all the inputs and advices shared here. I simply had no time to push further my testing at this time but it's planned as I'd like to understand better what's going on here. I fully second your ask for a wiki entry as well (I might use the existing "CPU Cycles" [entry ](https://github.com/dosbox-staging/dosbox-staging/wiki/CPU-Cycles-Howto) for that purpose.

All I've done thus far was to fiddle with the following settings. Unfortunately none did improve the situation. Below current settings I've changed based on your recommendations:
```
output = texture
core = dynamic
scaler = none
machine  = vesa_nolfb
```

Next I'll try to find the amount of cycles according to your explanations and I'll report back.

**Comment by kcgen** [comment]
Sounds good @nemo93 !

Also try avoiding per-frame OpenGL calculation overheard by setting `glshader = none`, and swap-in `output = texturenb` or `texturepp`, which only perform the calculation once on startup and then uses a hardware-scaling to stretch up to your TV size.

**Comment by nemo93** [comment]
I pushed further on this morning and got interesting (and strange) results. Please keep me honest and feel free to share more advices as I'm really intrigued by the outcome.

In short, I was able to go up to... **500.000 cycles** with the aforementioned process (start with fixed cycles amount and increase it by increment). Even at 500.000 cycles I was not able to saturate the CPU as I was still at 90% or below (85-90%). Tested with Quake 1 "timedemo" (with -nosound) and MDK.

The only issue is that the game(s) are playing at **slow-motion (and sound stutters)** with such "cycles" value. Therefore I decided to stick with Quake and go with a fixed value in the `.conf` file of the game ("cycles = fixed <value>"). To be sure.
=> It turns out Quake will start to "slo-mo" at around ~40.000 cycles or above. At 40.000 cycles the CPU is at 50% max or less (40-50%).
=> A good value for Quake to run is 35.000 cycles. With this amount the game is playable but you'll get below 20fps (around 15fps mostly). This cycles value equals a PI3B+...

MDK with "cycles = fixed 400000" will pass no problem the game's speed check with a score of roughly 300 which is way above the recommend (P200) score. Game is running but at slow-motion. 

Surprisingly even in slow-motion the games are "reponsive" meaning that whenever I press a key I got the correponding action on screen with no delay. I let both games to run for more than 10 minutes and everything was stable, no crash.

The settings from dosbox-staging-git.conf:
```
[sdl]
fullscreen       = true
display          = 0
fullresolution   = desktop
windowresolution = original
output           = texturenb (texturepp gave me black borders around screen)
texture_renderer = auto
capture_mouse    = seamless middlerelease
sensitivity      = 100
waitonerror      = true
priority         = higher,normal
mapperfile       = mapper-sdl2-git.map

[dosbox]
language = 
machine  = vesa_nolfb
captures = capture
memsize  = 16

[render]
frameskip = 0
aspect    = true
scaler    = none
glshader  = none

[cpu]
core      = dynamic
cputype   = auto
cycles    = auto
cycleup   = 5
cycledown = 20
```

To override the `[cpu]` settings I'm relying on one `.conf` per game. For Quake and MDK I've set:
```
[cpu]
cycles = fixed 500000 (max value I've been able to run with CPU ~90%)
cycleup   = 5
cycledown = 20
```

Note: with `output = texturenb` or `output = texturepp` TombRaider (yes again :)) will crash right after the Eidos logo as with Dosbox "SVN". With output = texture` it will crash but after the 2 FMV sequences and right before the main menu screen. I'll update the ticket accordingly.

**Comment by kcgen** [comment]
@nemo93 , nice digging, and interesting indeed.  

The way `output` mode changes Tomb Raider intro progress is very interesting, and probably worth of a separate bug report.  Maybe this can be reproduced on x86.  

Regarding the cycles and CPU usage - yes, I see this on x86 too (being able to crank way up to 1 M cycles).

Dosbox dedicates a lot of logic to cycle management and ultimately calls into SDL for its timer.  My gut feel is that there are subtle changes between SDL 1 and 2, and some of that cycle logic might need to be refined to take these into account.

**Comment by kcgen** [comment]
Another option to try is adding:

``` ini
force_turbo = 1
```

To `/boot/config.txt`. You will need sudo to edit it, and then reboot to take effect.

This will prevent your CPU from stepping down it's cpu speed, which might actually be happening on a microscopic scale during dosbox's 1ms forced-wait times.  Curious what you find!

**Comment by nemo93** [comment]
Adding `force_turbo = 1` to `config.txt` doesn't change anything to the previous results as far as my tests went. I need to redo another thorough testing session in the coming days. 

This really puzzles me. It's even more frustrating given I "feel" Pi4 got enough "power" to run more games fullspeed!

**Comment by kcgen** [comment]
Thanks for trying; that rules our aggressive power saving stuff on the Pi's side!

**Comment by nemo93** [comment]
FYI I just tried with the brand new 5.4 kernel (still in testing) and I haven't noticed any change on Pi4 in term of performance. Games are still running slower compared to Pi3.

**Comment by kcgen** [comment]
Oh man; that is just.. so tough to hear.   I'm a huge fan of dosbox on the Pi and was really hoping the Pi4 would give us enough "oomph" to reach up in the 60k -80k+ cycle range.

You've done everything possible to make sure it's not the Pi's fault (ie: thermal throttling or aggressive frequency idling); so maybe we're seeing the impact of poor memory locality causing rampant page-fault and cache eviction - so the improved CPU is being endlessly throttled by main memory speeds.

The Pi3 had DDR2 versus the Pi4's DDR4. DDR2 running at 400Mhz or (overclocked) to 500 Mhz+ in the Pi3 puts up pretty solid memory latency numbers versus slower DDR4-3200. https://en.wikipedia.org/wiki/CAS_latency

(1st, 4th, 8th latency):
DDR4-3200 CAS16 10.00 ns | 10.94 ns | 12.19 ns
DDR2 400Mhz CAS 4 | 10.00 ns | 13.75 ns | 18.75 ns
DDR2 533Mhz CAS 4 | 7.50 ns | 10.31 ns | 14.06 ns

For cache-unfriendly software, having a huge CPU cache helps hide the impact, but for CPUs without that then main memory latency is often the next throttle point.

You could measure the impact of this by underclocking your memory as far as possible, and then overclocking it as far as possible - and measure the span of dosbox's performance. (Really tiny, cache-friendly software will be minimally affected).
















https://en.wikipedia.org/wiki/CAS_latency

**Comment by dreamer** [comment]
A side comment from me: older RPis can have support for a subset of Vulkan API now (might be faster than OpenGL): https://www.phoronix.com/scan.php?page=news_item&px=RPi-VK-Driver

**Comment by nemo93** [comment]
@dreamer I saw this new driver and it just arrived after I've switched to Pi4 ;D Hopefully a version for the pi4 would come to life. When I see the result of some drivers with emulator like [redream](https://redream.io/) I just can't wait!

@kcgen At this stage I'm going to put on pause my "perf" testing as it goes nowhere and also because I don't want to spam you with it. FYI I've done another session with [lr-dosbox](https://github.com/libretro/dosbox-core) the results look better and in-line with what I got on Pi3. With fixed cycles of 40-50.000 it's ok with this version. It's also the only version where Tomb Raider is running flawless everytime no matter what.

**Comment by kcgen** [comment]
Thanks for the heads up regarding `lr-dosbox`  Nemo; I'll take a poke through it try to see how it differs.

**Comment by nemo93** [comment]
**UPDATE**: with the recent commits the perf on Pi4 are much better (starting from 6d087e488ccca03d89211ad909a81c1bcc32906b onwards). 

Games like Screamer(s) are playable (in VGA though), Fade to black, MDK and also the Build engine games - Blood, Duke3D, Redneck Rampage. 

However some other games still resist: apart from the already mentioned Tomb Raider random crashes, UAKM "slideshow", Alone in the dark serie, Crusader no remores/regret are still running very slow no matter what. Worth mentioning that some intros/animations are true "slideshows" (intro of Crusader no Remorse, the Psygnosis logo on Destruction Derby 2, Lemmings 3 intro, Tomb Raider fmvs, etc). Megarace 2 is not running any longer and I'm going to investigate as it worked before (crash right after the "please wait message" on launch).

I'll keep posting regularly with my findings.

**Comment by nemo93** [comment]
I feel a bit silly but I eventually found a way to run games at decent speed on pi4. In fact it's exactly what @kcgen suggested earlier. You need to be very specific with the `.conf` file settings of each game to run. Failing to do so or relying on default values will likely lead to crash, fmv like "slideshows", etc. The most common way to ensure games will properly run is **_to set a fixed amount of cycles_**. 

Some examples. By setting the default `cycles = auto (or max)` Alone in the Dark serie of games will suffer from very poor perf on pi4. If you set this value instead, games will run at fullspeed.
```
[cpu]
cycles = fixed 15000
```

Same goes with Crusader no Remorse/Regret. Leaving the default auto value will end up in intro fmv playing like a "slideshow" and loading times could take forever. By setting a fixed amount of cycles, those games will run at very decent speed:
```
[cpu]
cycles = fixed 15000
```

**Comment by kcgen** [comment]
Glad you're finding all the nuances to get acceptable performance from your Pi4 @nemo93 !

For future users, a one-stop-shop wiki article might do the trick (until we come up with an elegant way to control defaults for every platform). 

I was think you could expand this page to include all your recommended settings for both Pi models:

https://github.com/dosbox-staging/dosbox-staging/wiki/CPU-Cycles-Howto

Then rename the page title, ans it should feature prominiment i. The list of wiki articles.

**Comment by kcgen** [comment]
@nemo93 

I'm exploring some scheduler changes, and am how they influence overall performance and latency on your Pi4.

The branch is: `kc/fps-pacing-1`

Some quick points:
- It does not understand 'auto / max / %;, so give it a fixed number in your `conf`. For your Pi, I would say 50k up to 80k (although it might go higher). 
- You might want to pipe its output to a file `dosbox <game.exe> > /tmp/staging.log`, right now it's a bit noisy and the extra output to your framebuffer might slow things down.

The changes so far:
- It uses separate frame-timing thread paced off your processors "steady clock". 
- Frames used to be drawn in rapid succession at an uncontrolled rate followed (usually) by a one-second sleep period; kind of like getting all your work done on Monday and taking a break for the next 6 days.  The new code now draws only once every 16.666ms (yes, hardcoded right now), and your controls will be polled at the draw-time, so you might notice things are twitchier too.
- The PIC decode loop is now preempted if needed to to draw the frame (and keep latency low).
- It will catch when the things are about to bog down, and will "auto-throttle" the cycle count downward to keep latency beneath a single frame.  It will report this in the log message.
- At the same time, it will unthrottle when latency is tight again and re-add to the cycle count. (also reported).

If you try Quake, set your cycles to say 80k, and then check the log - it will show you when it's had to throttle down, but then it will try to re-gain your set point so long as it doesn't harm latency. 

On my desktop Linux box, I can lock the CPU to 800 MHz. When I play quake with `cycles = 200000` (which is quite a stretch for the CPU at 800MHz); I can see in the log that the sweet spot is around ~110k cycles for plenty of the busy play areas, however there are plenty of parts where cycles grow back up to ~150k and even 190k in simple hallways.  

Curious if any of this helps in any capacity!

**Comment by nemo93** [comment]
Let me thank you first for digging further, much appreciated. RPI4 really deserves better than the current ~30k or so cycles :-) I'll keep it short: after compiling and testing that branch the only improvement I was able to see was for Blood and Redneck Rampage. Game are running way smoother (at lowest resolution). With 80k cycles. It's worth noting that RR is "auto-throttling" like crazy: if you stand still you can hear cdda music properly however as soon as you start moving (or there's some action on-screen) music will slow down yet gameplay remains smooth. 

However the other Build-engine games don't run that smooth Duke3D and Shadow Warrior. Duke3D does run and it get stuck during the gameplay demo. Shadow Warrior is stuck on the very first Build screen where it states "load sound & graphics". After more than 5 real minutes I decided to quit.

Every other games I threw at this branch didn't run (I only tried 80k cycles). Well in fact they did launch but quickly dropped below the 500 cycles mark. It's even worse whenever there's a FMV. For instance Crusader No Remorse: with 80k cycles, game runs yet I can't barely see the very frst Origin logo. It took forever to display. Same goes with No Regret: if I'm quick enough I could pass through the logos/intro in order to reach main screen (pressing ESC like a crazy dude). However it's a pure slideshow and it takes forever to move from one option to the next. Same here on the cycles front: game almost always stays below the 500 cycles. Same stuff happened with Destruction Derby, Fallout, Command & Conquer, Dungeon Keeper crashed after the FMV intro and during the loading screen (red CD image).

Screamer gave me hope because this time it ran way better (still at 80k cycles). I was able to watch the very first demo race and main menu. However once the first option is selected music keeps on playing and cycles dropped to almost 0. With Screamer 2 I can't even see the demo race just the logos after that cycles count dropped.

Azrael Tears did run however I got stuck on the HUD display (bottom of the screen). Then cycles dropped to 0 and nothing further happened (you should see "hud boot sequence" as well as the 3D world but it's not the case here).

Quake at 80k appears to be just a bit better than with 35.000 cycles! At least during the very first demo. Gameplay looks a bit smoother but it's not what I expected. It's even worse when I start a game, it runs in fact slower and it turns into a slideshow whenever I get close to the Easy section and first level (left gate and then left door). I tried just for my personal education with 200k cycles and same thing occured.

In the log created I see the cycles dropping to almost 0 for the games aforementioned (except for Blood and RR).

**Comment by kcgen** [comment]
@nemo93 - thanks for the thorough testing; this made me recall how RetroPie's latency settings on the Pi also need to to be relaxed versus a typical desktop PC (`run_ahead_frames`, `video_max_swapchain_images`, and `video_frame_delay`).

I've updated the branch, and it now has a couple compile-time settings that let us relax these timings in a similar sense:

Top of **src/dosbox.cpp**:

``` c++
constexpr double FRAME_RATE = 60.0;
constexpr uint8_t MAX_FRAME_QUEUE = 9;
constexpr int MIN_CYCLES_PERCENT = 33;
```

- **MAX_FRAME_QUEUE** is how many frames worth of processing you're willing to queue (and accept, latency-wise), before the scheduler starts reducing the CPU cycles given to the DOS program in exchange to prevent more "debt" building up in the queue.

    In the version you tested above, the cut-off was a single frame -- and your tests prove that the Pi won't be able to maintain that for more CPU intensive games - even when the scheduler is kept cutting cycles all the way to zero. 

- **MIN_CYCLES_PERCENT** is how few cycles, as a percent of the fixed quantity specified in the `conf` file, that the scheduler can reduce to in order to improve latency.  

    In the version you tested above, the minimum was zero (and we see the Pi actually hit this; on my desktop, the lowest it went was ~60%).

The games you're testing are relatively cycle-hungry (granted, the Pi4 /should/ manage them if it's really 2x to 3x faster than the Pi3).  On my x86-64 desktop, these games need to be fed between 60k and 120k cycles to keep their latency below a single frame. 

I would start with the defaults above, and give the games the `cycles = 70000` range you tried before. You can try a deeper queue (say up to 15 or even 20 frames), and you could also bump the minimum percent to 50%, given we know the Pi4 should handle ~35k at a minimum.

**Comment by kcgen** [comment]
@nemo93 , I just pushed a small update (incase you pulled the branch before this message)

**Comment by kcgen** [comment]
Been testing on my Pi3 as well.
Here are some quick steps to overwrite your local branch with the latest changes:

``` shell
git fetch --all
git reset --hard origin/master
git reset --hard origin/kc/fps-pacing-1
make
```

**Comment by kcgen** [comment]
@nemo93 , on my Pi3B (not 3B+), I have the following cycles set with this branch:

**pi@retropie:~/dos $ grep cycles */*.conf**

``` text
Destruction Derby/dosbox.conf:cycles=40000
Duke3D/dosbox.conf:cycles = 40000
Quake/dosbox.conf:cycles = 45000
Redneck Rampage/dosbox.conf:cycles=fixed 40000
Screamer/dosbox.conf:cycles=fixed 40000
Shadow Warrior/dosbox.conf:cycles=fixed 50000
```

They all play pretty darn smooth (at the usual 320x200/240 resolutions), with tight control latency and no audio stutters. When I SSH in and run `top`, dosbox is extracting between 99%, 99.7, 100% CPU usage.

Compared to the master branch, I typically need to set cycles to 27000 - 30000 to prevent audio stutters, but then dosbox only extracts between 70 and 85%, and framerates is choppier. 

I sure hope the Pi4 does a bit better than this; but who knows.. it seems like quite a different beast.

Do I recall you mentioning having a 3B+ too? Would be great if you could confirm you're seeing similar performance atleast in the Pi3-realm (but only if that's easy for you to try).

**Comment by nemo93** [comment]
@kcgen - I'm very sorry but I gave away my good old 3B+ thus can't test with it any longer. I'm pi4-only now :-(

Thanks a lot for this new version. Lot better and I got very interesting results with all games. TL;DR: it's way better now.

Let's start first with the list of games above. I tested only with the defaults provided for now:

```
constexpr double FRAME_RATE = 60.0;
constexpr uint8_t MAX_FRAME_QUEUE = 9;
constexpr int MIN_CYCLES_PERCENT = 33;
```

* **Destruction Derby** / `cycles = 80000`
=> almost fullspeed I'd say. The only issue - in fact the main issue - is the "inconsistency" with music. I guess due to the auto-throttling sometimes music plays at normal speed and soon after "slows down" before playing faster again. My senses are confused :-) what my eyes see on-screen appears to be smooth whereas my ears disagree!

* **Duke3D** / `cycles = 80000`
=> almost fullspeed. Sometimes there are slow downs like in the main menu transition while game loads demo in background. After few seconds though the fps number stabilizes.

* **Quake** / `cycles = 70-80000`
=> still not smooth. I've tried 70 and 75k as well and same results. I know Quake was never meant to be 60fps and with this branch results are definitely playable yet it's not super smooth (I'm comparing to the Quake Raspberry port although I know it will be hard to get the same fps number). For instance during the first demo when the grenade hits the first group of zombies you have almost a pause in the gameplay. No more slideshow though. In log I found out the 70-80000 limit is never being reached. Scheduler stays in the 35-45000 cycles.

* **Redneck Rampage** / `cycles = 70000`
=> for some reason it was smoother yesterday! Duke3D and Shadow Warrior are lot smoother. I'll try with 80000. It's very playable though.

* **Screamer** / `cycles = 70000`
=> note: I'm running with the `STARTL.EXE` binary. I might give a try to the `STARTH.EXE` too. I'd say game is fullspeed. Very smooth.

**Shadow Warrior** / `cycles = 75000`
=> some slow downs yet it's way smoother than ever before.

I've tested with many other games mostly using the `70-80000` cycles zone, cpu usage on pi4 never went above 60-70%. Below some outstanding results I made notes of in no specific order.

* **Azrael Tears**: `cycles = 80000`. I **got a crash** with that one. First time I got this. Too bad because this time the performance achieved are really nice. It's not super smooth but game is definitely playable which is some kind of achievement given how demanding this game is. No log was generated unfortunately.
* **MDK**: `cycles = 80000`. No more screen at launch stating CPU is not powerful enough. Performance counter from the options menu assigned a solid 35 which is not bad. It allows the game to be playable definitely.
* **Command & Conquer**: I went w/ `cycles = 60000` and no more slideshows during FMV. Game is playable. The only issue worth noting is a heavy slowdown (almost a pause) anytime there's an audio track change.
* **Destruction Derby 2**: `cycles = 80000`. This one result is impressive as game is almost fullspeed! Even with the draw distance pushed to the max game is running absolutely flawless.
* **Theme Hospital**: `cycles = 80000`. Whenever there are lots of patients game starts to slowdown. I recall it was the case back in the days but I was expecting better. I'm playing with shadows on and in high resolution. Playable though.
* **Dungeon Keeper**: `cycles = 70000`. Exactly as I recall back in the days! Gameplay is now super smooth (so does the FMV). The only issue I'd say is during the map sequence preceding each level. The zoom in/out remains slow.
* **Screamer 2**: `cycles = 80000`. Very smooth with the `START65L.EXE` version.
* **Fatal Racing/Whiplash**: `cycles = 80000` is now playable with almost all details on (in VGA)!
* **Fade to Black**: `cycles = 70000`. At "medium" resolution game is now fullspeed. The high resolution is ok-ish but too slow in my opinion to be considered playable.
* **Crusader No Remorse/Regret**: `cycles = 70/80000` are still not smooth. There are slowdowns (especially w/ Regret) and I'd 0.5-1s pause delay whenever there's a screen change (which happens a lot!). FMV are fullspeed now including the Origin logo on Regret which never played properly before.
* **Death Rally**: `cycles = 70000`. It's playable yet it suffers from the slow/fast speed issue mentioned above. It can be annoying as sometime your car slowdowns before being propelled forward the next second.
* **Tomb Raider**: `cycles = 70000`. Almost fullspeed using the high details setting! No more crash or slideshows definitely :)
* **Wipeout**: `cycles = 75000`. Gameplay wise it's almost fullspeed this time. Only issue is the same as with Death Rally and Destruction Derby: you got music slowing down sometimes (right from the main menu).
* **Blood**: `cycles = 80000`. As with Duke3D you got an initial slowdown during the demo/main menu but during normal gameplay it's running at very good speed. Definitely playable.

Side notes:
1. I still can't get **Ecstatica 2** running. Game does launch but got stuck right after the Ecstatica 2 logo where you got a screen with World and Design by... on it. Game got stuck there. I've been using `cycles = 80000`.

2. In Redneck Rampage 2 (Rides Again) log I saw the following line:
`CDROM: GetTrack at sector 232205 is outside the playable range`

That's all for today  \o/

**Comment by kcgen** [comment]
We're making some real progress @nemo93!  Would you mind if I made you co-author on these commits? Your detailed exploratory work and feedback has been critical in moving the scheduler improvements ahead.

- Of that games that were improved - would you prefer using this over other options - has it notably improved your experience  (versus SVN, dosbox-staging master or 0.75.x, or DOSBox-X.  I just want to be sure that overall, this is still a win - and we should keep going.  

- If we're still not doing better than SVN or -X, would you say this is still an improvement over the baseline dosbox-staging?

- Thanks for the all the show-stoppers and undesirable behavior (doppler CDDA in Destruction Derby); these can be fixed by tightening reducing the max latency (back it down to one or two frames), and giving the scheduler less leeway to reduce cycles (or just hold it fixed, like default) - all of which is pointing to making these knobs adjustable in the conf file; something like:

```
[scheduler]
max_latency = 7
reduction_max_percent = 50
reduction_rate = fast  
recovery_rate = slow 
```

I'm not a fan of adding conf options - so if we did go with something like this, we'd use very sane defaults; but what we're seeing is that each game might have an optimal scheduler tuning. 

For example, these heavier games like Quake and the Build Engine games really become playable when they're given more latency leg room, like you mentioned .. (before, I didn't even attempt these on the Pi3.. they were just 5fps slide shows).

For those DOS game preservation, the goal would be to find the optimal settings for a given game - especially the troublesome games; and lock these in for a given game.

**Comment by nemo93** [comment]
hey @kcgen I fully agree. Definitely you're up to something good. In just a couple of days - thanks to you - I've seen more progress on the performance front for RPi than over the past couple of years! This without even mentioning all the improvements implemented and about-to-be implemented. Simply awesome. 

> Would you mind if I made you co-author on these commits? Your detailed exploratory work and feedback has been critical in moving the scheduler improvements ahead.

=> you're doing all the hard work :) I'll be glad to be put as co-author but it's really your work that's important.

> Of the games that were improved - would you say this is an overall improvement versus SVN, dosbox-staging master (or 0.75.x), or DOSBox-X? I just want to be sure that overall, this is a win and we should keep going.

=> Definitely an overall improvement vs SVN and master. This branch (`kc/fps-pacing-1`) is my new default given how well it performs. And I haven't even fiddled with the scheduler parameters! I don't have DOSbox-X to compare with. I tried this fork once but the massive amount of options and features had me to step back. It was at that time I discovered dosbox-staging :) Since then I stick to it!!

> For some of these games, I noticed you can really punch up the cycles, and things really get fluid in the gameplay; like if you tried between 120000 or even 140000 (for example, when I locked my desktop to 800 MHz, I can go up to 150k in Screamer 2 which makes the high-res version very fluid; on the Pi3, I can push 50k and it makes the low-res build engine games very smooth too).

=> correct. It was my plan to do several passes of testing. Since you mentionned 80k and 120k I decided to try first the lower value with my set of games candidates. Today I've pushed further. The max cycles count I've been able to reach was `200000`. This number is more an exception than a rule. I'd say more games would accept a ~150k cycles without any stuttering or sound glitches. I won't go to another detailed list I'm just highlighting the most interesting ones.

_Pi4 never went above 80% of cpu usage during testing. Even at the highest count cycles._

* **Fade to black**: `cycles = 200000`. With this setting the game is fully playable using the highest resolution! I was really impressed. Also I'm happy as I have fond memories of that one.
* **MDK**: `cycles = 150000`. No more screen at launch stating CPU is not powerful enough. Performance counter from the options menu assigned a solid 35-45 which is not bad. It allows the game to be playable definitely. I don't see much difference between the low/high resolution setting from the options though. I'd say it's barely playable.
* **Crusader No Remorse/Regret**: `cycles = 60000`. Setting a higher number will lead to slowdown. Also for the pauses between 2 screens I mentioned earlier it seems that `machine = svga_s3` is helping a lot. Both games are fully playable but it's not perfect.
* **Destruction Derby 2**: `cycles = 200000`. Full speed in low resolution mode! No "doppler" effect with music. High resolution is not playable.
* **Blood**: `cycles = 200000`. Full speed! It's the game which runs best amidst all Build engine games.
* **Duke3D/Shadow Warrior**: `cycles = 130000`. From the internal fps counter I stick between 45 and 60fps.
* **Death Rally**: `cycles = 70000`. Can't go above that number. "Doppler" effect and slowdowns to be noted. I'd say this one is the most disapointing.
* **Fatal Racing/Whiplash**: `cycles = 100000`. Playable in VGA mode! Increasing the cycles count will generate stuttering.
* **Dungeon Keeper**: `cycles = 100000`. Fullspeed to me. Seems that `machine = svga_s3` helps with the zoom in/out slowdown.

* **Screamers** games are still not running properly using the highest resolution setting. They're both fine only when using the "low" resolution setting (65L for Screamer2). I can't go higher than `cycles = 100000` or stuttering will appear.

> Thanks for the all the show-stoppers and undesirable behavior (doppler CDDA in Destruction Derby); these should be fixable by tightening the max latency and/or giving the scheduler less cycle-leeway (but I need to test them); I think all of this is pointing to making these knobs adjustable in the conf file

=> I'm not a fan either of having a plethora of options/settings. Yet given the overall improvement seen here I'll be fine having them somewhere in the conf! Without even changing any of these scheduler settings results are quite impressive already. Simply by tweaking the cycles count. If users want to start fine-tuning they would do so thanks to those scheduler options.

Let me know if you come up with something else. I'll be happy to test. Thanks!

**Comment by kcgen** [comment]
@nemo93 , thanks for the very positive feedback! 
It's fascinating that these small timing and preemptive tweaks can influence performance so much.

Even on my Pi3.. seeing some heavy late 90s games now running butter smooth tells me dosbox has a fantastic engine at it heart- the dynamic recompiler, and the emulated devices running natively on the host side. It just needs a different gearbox on some platforms.

Wanted to say that this is going to be the next task in my queue once 0.75.1 and 0.76 are on their way, and something we can really polish up. 

 I will start by adding latency and variance measurements for input (keyb/mouse/joy) polling and frame drawing, so we can measure the current scheduler and then objectively compare it against the preemptive pacing scheduler.

**Comment by arrowgent** [comment]
what are some general considerations with the compiler?
CLANG vs GCC?
clang creates a 2.8MB binary, where as GCC creates a 3.4MB binary
what differences in speed could be considered here?
or is this mostly a consideration in how fast the binary loads?

i compiled dosbox-staging on my RPI4B for testing

**Comment by kcgen** [comment]
@arrowgent , thanks for jumping in and testing on your Pi4.  GCC almost always produces a faster-performing executable, so you will want to build with `./scripts/build.sh -c gcc -t release`

But if your coding and debugging, clang can provide better messaging and warnings, so I use  `./scripts/build.sh -c clang -v 10 -t warnmore` for that. (The -v 10 targets my clang 10 binaries). 

Executable size and startup time are not metrics were care too much about; runtime performance and getting smoother gameplay on the Pi (especially on the heavy games @nemo93 flagged above) is the primary goal.

**Comment by arrowgent** [comment]
some "cheat codes" to show FPS for build games:

duke3d: type `dnrate` during game shows FPS rate
swarrior: press T to type `swloc` shows FPS rate
blood: press T to type `rate` shows FPS rate

taking note that originally 35fps was the default desired rate for all of the build engine games.

for raspberry i set all resolutions to lowest 320x200 for these.  (mostly use dosbox as an ipx server and testing)

**Comment by arrowgent** [comment]
gcc compiled 0.75.1
raspberry 4b 4GB rev 1.1
`arm_freq=1866 (clock 1.9GHZ)`
`gpu_freq=500 (default)`
other minor tweaks
* "52pi" heatsink
* https://www.amazon.com/GeeekPi-Raspberry-Low-Profile-Cooling-Heatsink/dp/B07ZV1LLWK/
with noctura fan
* https://www.amazon.com/Noctua-Cooling-Blades-Bearing-NF-A4x10/dp/B009NQLT0M/
keeps cool 35~40C max CPU%

as noted here i setup everything for low-end basic specs:
```
[sdl]
fullscreen       = false
display          = 0
fullresolution   = 1920x1080
windowresolution = 1280x960
output           = texturenb
texture_renderer = opengles2

[dosbox]
machine  = vesa_nolfb
memsize  = 31

[render]
aspect    = true
scaler    = none
glshader  = none

[cpu]
core      = dynamic
cputype   = auto
cycles    = fixed 100000

[mixer]
nosound   = false
rate      = 22050

[midi]
mpu401     = none
mididevice = none
midiconfig = 

[sblaster]
sbtype  = sbpro2
sbbase  = 220
irq     = 7
dma     = 1
hdma    = 5
sbmixer = true
oplmode = auto
oplemu  = fast
oplrate = 22050

[gus]
gus      = false

[speaker]
pcspeaker = false
pcrate    = 44100
tandy     = off
tandyrate = 44100
disney    = false

[joystick]
joysticktype  = none


[serial]
serial1       = disabled
serial2       = disabled
serial3       = disabled
serial4       = disabled

[ipx]
# ipx: Enable ipx over UDP/IP emulation.
ipx = false

[autoexec]
# Lines in this section will be run at startup.
# You can put your MOUNT lines here.
@mount c ~/Games/dosbox
@c:
```
i dont actually use sound, but since "emulation cannot be disabled" i left it on basic configuration

some testing monitoring `HTOP` from `C:\>`
only modifying
```
cycles    = fixed 100000
```
with `config -set "cpu cycles=fixed #` from inside dosbox
cycles 1000 = 13-15%cpu
cycles 5000 = 16-17%cpu
cycles 10000 = 22-24%cpu
cycles 20000 = 27-31%cpu
cycles 30000 = 29-32%cpu
cycles 40000 = 36-41%cpu
cycles 50000 = 40-44%cpu
cycles 75000 = 58-62%cpu
cycles 100000 = 72-75%cpu
cycles 130000 = 89-94%cpu
cycles 140000 = 95-98%cpu
cycles 150000 = 95-97%cpu

from this i can see that 140,000 is around the peak cycles for my setup.

testing in game: BLOOD
start game, stand in grave dont jump: 
1) walk forward observe stable fps
2) walk backwards obserbe stable fps

forward cycles 5000 = 2fps
backward cycles 5000 = 2fps
forward cycles 5000 = 2fps
backward cycles 5000 = 2fps
forward cycles 5000 = 2fps
backward cycles 5000 = 2fps

**Comment by kcgen** [comment]
That's a really nice Pi setup you've got @arrowgent; and getting 1.9Ghz from it's default 1.5Ghz CPU speed is an impressive overclock!

The default dosbox scheduler will discards it's work queue when it's more than 20ms "in the hole", which is why audio will stutter and dropped frames we set cycles too high.  The game itself doesn't realize this though, so (usually) FPS counters will just keep going up and reporting better and better "performance".

But yeah - like you've done, the real edge is when you can see and feel the latency or audio-chop.

**Comment by arrowgent** [comment]
> That's a really nice Pi setup you've got @arrowgent; and getting 1.9Ghz from it's default 1.5Ghz CPU speed is an impressive overclock!
> 
> The default dosbox scheduler will discards it's work queue when it's more than 20ms "in the hole", which is why audio will stutter and dropped frames we set cycles too high. The game itself doesn't realize this though, so (usually) FPS counters will just keep going up and reporting better and better "performance".
> 
> But yeah - like you've done, the real edge is when you can see and feel the latency or audio-chop.

ill enable audio and check some things.
i have other dos games on linux, but only Build games are on my RPI
im working on GOG Carmageddon multiplayer, so i could pull that over and test that.

as for overclock, i tested using 2GHZ but it became unstable after about 3-7 days and crashed
i also underclock the `arm_freq_min=333` default is 500 so it does give some different c_state (or p_state) dividers

**Comment by kcgen** [comment]
> i also underclock the arm_freq_min=333 default is 500 so it does give some different c_state (or p_state) dividers

Cool.. I didn't realize you could underclock on the slow-side. I'll have to try something similar on my Pi3.

**Comment by arrowgent** [comment]
with audio on
SFX = soundblaster
MIDI = soundblaster

cycles = 175000 very choppy audio
cycles = 150000 less choppy
cycles = 140000, 130000, 120000, 110000 still choppy
cycles = 100000 nearly tolerable but has pops and waves
cycles = 90000 almost clear enough
cycles = 80000 mostly clear,  im seeing the CPU spike up to 96%
cycles = 70000 clear audio but CPU spikes 96-98%
cycles = 60000 looks and sounds good, cpu between 88-90% during some spikes


since my resolution is 1280x960 windowed, anyone else might experience different results in different Window sizes
i dont prefer to play fullscreen.

**Comment by arrowgent** [comment]
it might be notable to test
```
oplemu  = fast
```
for the other options and see if cycles effects this dramatically

**Comment by kcgen** [comment]
On Pi3 and Pi4, if you apply the SDL2 patch to DOSBox, you will experience similar slows downs reported here.  I personally experienced this in ~2017 when I attempted exactly that using NY's patch on Vogons on my Pi3.

But these reports continue today:

- https://retropie.org.uk/forum/topic/24876/rpi4-slow-and-laggy-dosbox-sdl2-version

> Running the dosbox-sdl2 version is laggy and slow on occasion. I'm running a weekly retropie buster build on a 4gb RPI4. Example game; Command and Conquer Maybe, just maybe this is vsync/sdl2 related?

This broadens to SDL2 in general:

https://retropie.org.uk/forum/topic/24689/frameskip-issues-on-pi4

> Many of you have certainly experience a strange frame skipping on Pi4 that wasn't present in the previous incarnations. This is usually manifest in the form of "losing every second frame", consistently, similarly to running 60hz signal on a 30hz rendering. 

> This is caused by the new graphic drivers required to run the Pi4 (new chipset, new drivers). The issue happens because of a bug/lack of support in the new drivers. You can find more info here.

> Several emulators and ports are affected:
> Drastic
>  Prince of Persia Port
>  Oric
>  VVVVV
>  etc.

A reference to a fixed version is mentioned here:

- https://retropie.org.uk/forum/topic/24688/solved-rpi4-retropie-weekly-build-scummvm-vs-lr-scummvm-slow-graphics

The RetroPie maintainer bumps their setup script to put a newer internal-build of SDL2 for the Pi4 here (this isn't a repo-maintained SDL2, it's RetroPie's custom-built version):

- https://github.com/RetroPie/RetroPie-Setup/commit/7c74cff0cf47306f8bdec2c4bb1c6af221a5fbb6

The comment mentions reverting a KMS non-vsync issue, and the closest match I could find in SDL2's commit history is this: 

https://hg.libsdl.org/SDL/rev/33bcd45f0447

So perhaps we have more avenues to further improve performance by trying a custom builds of SDL2, or waiting for these fixes (and/or to the RPi4 graphics drivers) to land in Raspbian's repo.

**Comment by nemo93** [comment]
Just for the reference I also do have high hopes for better performance on pi4 thanks to the new 5.10 kernel + Full KMS driver + 64-bit:

https://retropie.org.uk/forum/topic/25370/good-news-regarding-tearing-and-low-emulationstation-performance-on-pi-4

**Comment by kcgen** [comment]
This is _very_ exciting: finally, a native kernel mode-setting driver and not a lag-inducing software wrapper. 
I now consider the Pi4 existing it's beta stage and have placed an order :-) 
@nemo93 - thanks for keeping such a close watch on this! :100:

**Comment by grapeli** [comment]
I ran dosbox-staging with sdl2 kmsdrm output (opengl) on an old laptop with IntelHD Gen 5.5th. I must admit that this output (kmsdrm) is by far the fastest. About 5-9% compared to x11 or wayland. Although on a "desktop", unfortunately a bit of little use.

**Comment by kcgen** [comment]
@nemo93 , a user on discord recorded the following Tomb Raider performance on their Pi4 (using the eXoDOS v5 configuration of the game). 

They were quite happy with the 320x200 resolution performance using `machine = vgaonly`:  https://streamable.com/88o6la

As the default `machine = svga_s3`: https://streamable.com/xptcay performs worse. 

Are we competitive with either of these?

**Comment by dreamer** [comment]
Tip (because it's not obvious at all): Tomb Raider I starts in resolution 640x480 by default  - to switch to 320x240 in-game, press F1.

**Comment by nemo93** [comment]
@kcgen @dreamer Tomb Raider is now running pretty fine with `cycles = 100000` here on pi4 with high details, full screen size and 320x200 resolution. It's as close from fullspeed as I recall so this is really great. I do confirm the following for Tomb Raider:

* using `machine = vgaonly` is not recommended as you won't see the FMV intros and get a messy home screen. I stick to `machine = vesa_nolfb` without any issue.
* `cycles = 100000` with the now famous `kc/fps-pacing-1` branch gives best performance on pi.
* no specific other settings in the `.conf `apart `timed = true` for joystick to properly work.
* **F1** indeed will switch resolution (320x240 or 640x480), **F2** will reduce screen size while **F3** will increase it and **F4** switches between level of details (low, medium, high).

Frankly, I don't see any further improvement that could be done for that game. It runs _flawless_ on pi4. Perhaps 3DFX at some point but it's a bit early to speak about it :D

Thanks both for your continuous support and feedback.

**Comment by nemo93** [comment]
@kcgen @dreamer hello! Hope you're doing great. It's been a while since I've played with Staging on my pi4. I'm still carefully following the many commits merged on a daily basis and I decided to give a try to the latest master...

The difference with "older" releases is **spectacular**.

I'm able to play any games I could think of, ***fullspeed***. This on a simple stock pi4 \o/ No KMS or 64bit here, just the good old device in all its glory.

To give you an idea of such achievement, Quake is now running fullspeed (I need to get some accurate data but we're talking 60fps there and 30fps using higher resolutions), same goes with Wing Commander 4 (SVGA!), Destruction Derby's, Screamer's, Dark Forces, Crusader no Remorse/Regret, Duke3D, Blood or Doom/2.

Truly impressive. I'll run into another extensive testing session to be sure it's not an illusion :-)

There's the **new vsync parameter that must be set to true** in `dosbox-staging.conf` otherwise the display is not properly "refreshed" and heavy tearing might occur in-game. 
=> for instance, running Doom with `vsync = false` (default) then you'll get straight to a black screen instead of the menu (you can still hear sound fx/music in the back). With Crusader no Remorse only half the menu screen will be shown and if you circle through the menu options, only 1 out of 2 will be selected. With `vsync = true` no more issues.

To share some early data, for the most demanding games (Doom, Blood/Build engine games, Quake), I'm now able to increase the `cycles` count up to ~**70-80.000** without issue. Games are running at ~60fps.

I've posted a similar post on Retropie's forum: https://retropie.org.uk/forum/topic/25041/dosbox-official-thread/128 as I'd like to know if the outcome is the same on pi3.

**Comment by nemo93** [comment]
@kcgen perhaps your famous and very helpful thus far branch "fps-pacing" is no longer required? Given the huge performance boost on master I'm sticking to it from now on unless you tell me otherwise.

@kklobe thanks a lot as well for all the addition and that big performance improvement!

**Comment by kklobe** [comment]
@nemo93 thanks for stopping by and giving us an update on the performance. We're happy to hear all the changes have helped out.

Please keep us posted if you find new issues!

**Comment by nemo93** [comment]
Given this breakthrough and because performance now on pi4 are flawless I'm closing this one off. Big thanks to the team!

**Comment by symbios24** [comment]
Hello there

I managed to run Windows 98 in Dosbox-X in a Raspberry Pi 4,but the performance in games only is very slow, Desktop performance is ok,i tried various settings in config file without any speed improvement,i saw videos in youtube with raspberry pi and Dosbox svn/ece running games like age of empires 2 in full speed

https://m.youtube.com/watch?v=lYmfklwWK2k

https://m.youtube.com/watch?v=W5vt6jtzhlg

But i could not install windows 95/98 in dosbox svn and ece(getting errors on first boot,explorer.exe/dll crashing) is there a tip so games run normal or i can manage to install windows 95/98 in dosbox svn/ece without errors ??

I tried dosbox-staging(installed from retropie scripts) but upon running the keyboard layout was totally wrong for example if i press to type:a word it was typing symbols like numbers or totally not what i type.so i could not try windows from there.

Ps my pi4 is running at 1800mhz

**Comment by johnnovak** [comment]
DOSBox Staging doesn't currently support any Windows version above Win 3.x. From the "non-goals" section our about page here:

https://dosbox-staging.github.io/about/#non-goals

> Support the use of Windows 9x/Me in the emulator. Windows 9x/Me emulation is supported by projects such as [QEMU](https://www.qemu.org/) and [VirtualBox](https://www.virtualbox.org/), although DOSBox forks like [DOSBox-X](https://www.dosbox-x.com/) and [DOSBox Pure](https://github.com/schellingb/dosbox-pure) also support this. This may change if wanted by the DOSBox Staging community in general.

Some people are experimenting with Win95/98 stuff on Staging for fun, but as I said, it's not among our goals and we don't support it, don't fix Win95/98 bugs, etc.

**Comment by symbios24** [comment]
> DOSBox Staging doesn't currently support any Windows version above Win 3.x. From the "non-goals" section our about page here:
> 
> https://dosbox-staging.github.io/about/#non-goals
> 
> > Support the use of Windows 9x/Me in the emulator. Windows 9x/Me emulation is supported by projects such as [QEMU](https://www.qemu.org/) and [VirtualBox](https://www.virtualbox.org/), although DOSBox forks like [DOSBox-X](https://www.dosbox-x.com/) and [DOSBox Pure](https://github.com/schellingb/dosbox-pure) also support this. This may change if wanted by the DOSBox Staging community in general.
> 
> Some people are experimenting with Win95/98 stuff on Staging for fun, but as I said, it's not among our goals and we don't support it, don't fix Win95/98 bugs, etc.
> 
> My advice: use PCem instead.

I tried Ocemu but could not compile it i get an error after make

Do you know how to compile latest version 17 for Raspberry Pi 4??

Thank you

**Comment by rderooy** [comment]
This is not the place to ask about compiling other projects. Also, I would suggest you use 86Box, as unlike PCem, is actually actively maintained. In my limited testing with it, it was (on my AMD desktop) faster than DOSBox-X in regular use.

**Comment by symbios24** [comment]
I tried 86xbox but could not compile it either

Anyway thanks
